cmd/sdkstats/main.go

```diff
@@ -7,46 +7,120 @@ import (
 )
 
+type projectVersionParser struct{}
+
 func main() {
 	var token string
 	flag.StringVar(&token, "token", "", "GitHub API token")
 	flag.Parse()
-	args := flag.Args()
+	if len(token) == 0 {
+		log.Fatal("GITHUB API TOKEN MUST BE ENTERED \n Usage: './main --token=YOURTOKEN'")
+	}
 	ctx := context.Background()
 	ts := oauth2.StaticTokenSource(
-		&oauth2.Token{AccessToken: args[0]},
+		&oauth2.Token{AccessToken: token},
 	)
 	tc := oauth2.NewClient(ctx, ts)
 	client := github.NewClient(tc)
 
-	var queries []string
-
-	queries = append(queries, "filename:Dockerfile quay.io/operator-framework/helm-operator")
-	queries = append(queries, "filename:Dockerfile quay.io/operator-framework/ansible-operator")
-	queries = append(queries, "filename:go.mod github.com/operator-framework/operator-sdk")
-	queries = append(queries, "filename:Gopkg.toml github.com/operator-framework/operator-sdk")
-
+	queries := []sdkstats.RepoMetadataQuery{
+		sdkstats.RepoMetadataQuery{
+			ProjectType:   "ansible",
+			Queries:       []string{"filename:Dockerfile quay.io/operator-framework/ansible-operator"},
+			VersionParser: &projectVersionParser{},
+		},
+		sdkstats.RepoMetadataQuery{
+			ProjectType:   "helm",
+			Queries:       []string{"filename:Dockerfile quay.io/operator-framework/helm-operator"},
+			VersionParser: &projectVersionParser{},
+		},
+		sdkstats.RepoMetadataQuery{
+			ProjectType: "go",
+			Queries: []string{
+				"filename:go.mod github.com/operator-framework/operator-sdk",
+				"filename:Gopkg.toml github.com/operator-framework/operator-sdk",
+			},
+			VersionParser: &projectVersionParser{},
+		},
+	}
+	//GetStats function for Query String from 'queries', These Strings are specific to Operator-SDK patterns.
 	for _, r := range queries {
 		stats, err := sdkstats.GetStats(client, r)
-		if len(stats) == 0 {
-			log.Fatal("GITHUB API is down at time, Please try after few minutes")
-		} else {
-			fileName := stats[0].ProjectType + ".json"
-			if err != nil {
-				fmt.Println(err)
-			}
-			fmt.Println("Total count: ", len(stats))
-			file, _ := json.MarshalIndent(stats, "", " ")
-			_ = ioutil.WriteFile(fileName, file, 0644)
-			fmt.Println("Results are written in ", fileName)
+		if _, ok := err.(*github.AcceptedError); ok {
+			log.Println("Job is scheduled on GitHub side")
+		} else if _, ok := err.(*github.RateLimitError); ok {
+			log.Println("Rate Limit has reached.")
+		} else if err != nil {
+			fmt.Println("Failed to get Stats for Query", r, err)
 		}
+		//Write results into JSON , named with ProjectType.
+		fileName := r.ProjectType + ".json"
+		fmt.Println("Total count: ", len(stats))
+		file, _ := json.MarshalIndent(stats, "", " ")
+		_ = ioutil.WriteFile(fileName, file, 0644)
+		fmt.Println("Results are written in ", fileName)
 	}
 
 }
+
+//Parse the given Code result to search Text Matches for Version number.
+func (p projectVersionParser) ParseVersion(codeResults github.CodeResult, projectType string) (string, error) {
```

Comment:
I am not sure., I completely understand this. Since "Gopkg.toml" results , would not need ParseVersion, as they do not have defined search Query, I am defaulting sdkVersion to "N/A". Hence, I thought its unnecessary to make function calls to ParseVersion from sdkstats.go for below condition , 		if strings.Contains(q, "Gopkg.toml") {
			skipVersion = 1
		}

**"I don't think you need projectType in the function signature, since you're passing an implementation of this into GetStats as part of a RepoMetadataQuery that already includes the project type."**
So, You are saying its better to have separate ParseFunction defined, rather than passing "ProjectType", from sdkstats? Or May be I use which is available in Main.go, Not sure if that thread will still give me  current ProjectType value passed into sdkstats.go.

Reply:
Think of it from a library perspective and pretend you don't know what queries other users will be sending in. In that case, all the library can do is use the provided version parser.

So then from our perspective in `main.go`, think of things in terms of the API we have available, which is the VersionParser interface. In the case of `Gopkg.toml`, we're unable to parse a version so we define a version parser whose sole job is to return `"unknown", nil`.

```go
type unknownVersionParser struct{}
func (_ unknownVersionParser) ParseVersion(_ github.CodeResult) (string, error) {
	return "unknown", nil
}
```

Since that means we have different implementations for parsing go.mod and Gopkg.toml, it probably means we need to split our `ProjectType: "go"` queries into two `ProjectType: "go"` queries like this:

```go
queries := []sdkstats.RepoMetadataQuery{
	sdkstats.RepoMetadataQuery{
		ProjectType:   "ansible",
		Queries:       []string{"filename:Dockerfile quay.io/operator-framework/ansible-operator"},
		VersionParser: &baseImageVersionParser{
			baseImage: "quay.io/operator-framework/ansible-operator",
		},
	},
	sdkstats.RepoMetadataQuery{
		ProjectType:   "helm",
		Queries:       []string{"filename:Dockerfile quay.io/operator-framework/helm-operator"},
		VersionParser: &baseImageVersionParser{
			baseImage: "quay.io/operator-framework/helm-operator",
		},
	},
	sdkstats.RepoMetadataQuery{
		ProjectType: "go",
		Queries: []string{
			"filename:go.mod github.com/operator-framework/operator-sdk",
		},
		VersionParser: &goModVersionParser{
			module: "github.com/operator-framework/operator-sdk",
		},
	},
	sdkstats.RepoMetadataQuery{
		ProjectType: "go",
		Queries: []string{
			"filename:Gopkg.toml github.com/operator-framework/operator-sdk",
		},
		VersionParser: &unknownVersionParser{},
	},
}
```

Reply:
I agree with separating the  queries for "go.mod", and "Gopkg.toml". I was having them as separate queries in the beginning as well, to make VersionParse easier. Will make the changes.